mllib/src/main/scala/org/apache/spark/mllib/stat/test/KSTest.scala

```diff
@@ -0,0 +1,191 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.stat.test
+
+import org.apache.commons.math3.distribution.{NormalDistribution, RealDistribution}
+import org.apache.commons.math3.stat.inference.KolmogorovSmirnovTest
+
+import org.apache.spark.rdd.RDD
+
+/**
+ * Conduct the two-sided Kolmogorov Smirnov test for data sampled from a
+ * continuous distribution. By comparing the largest difference between the empirical cumulative
+ * distribution of the sample data and the theoretical distribution we can provide a test for the
+ * the null hypothesis that the sample data comes from that theoretical distribution.
+ * For more information on KS Test: https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test
+ *
+ * Implementation note: We seek to implement the KS test with a minimal number of distributed
+ * passes. We sort the RDD, and then perform the following operations on a per-partition basis:
+ * calculate an empirical cumulative distribution value for each observation, and a theoretical
+ * cumulative distribution value. We know the latter to be correct, while the former will be off by
+ * a constant (how large the constant is depends on how many values precede it in other partitions).
+ * However, given that this constant simply shifts the ECDF upwards, but doesn't change its shape,
+ * and furthermore, that constant is the same within a given partition, we can pick 2 values
+ * in each partition that can potentially resolve to the largest global distance. Namely, we
```

Comment:
Credit goes to @syrza for the algorithm

In mllib/src/main/scala/org/apache/spark/mllib/stat/test/KSTest.scala
<https://github.com/apache/spark/pull/6994#discussion_r33717824>:

> +
> +/**
> + * Conduct the two-sided Kolmogorov Smirnov test for data sampled from a
> + * continuous distribution. By comparing the largest difference between the empirical cumulative
> + * distribution of the sample data and the theoretical distribution we can provide a test for the
> + * the null hypothesis that the sample data comes from that theoretical distribution.
> + * For more information on KS Test: https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test
> + *
> + * Implementation note: We seek to implement the KS test with a minimal number of distributed
> + * passes. We sort the RDD, and then perform the following operations on a per-partition basis:
> + * calculate an empirical cumulative distribution value for each observation, and a theoretical
> + * cumulative distribution value. We know the latter to be correct, while the former will be off by
> + * a constant (how large the constant is depends on how many values precede it in other partitions).
> + * However, given that this constant simply shifts the ECDF upwards, but doesn't change its shape,
> + * and furthermore, that constant is the same within a given partition, we can pick 2 values
> + * in each partition that can potentially resolve to the largest global distance. Namely, we

I like the algorithm:) Is there a reference, or is it original?

—
Reply to this email directly or view it on GitHub
<https://github.com/apache/spark/pull/6994/files#r33717824>.